app/src/main/java/ru/spbhse/pocketmagic/Controller.java

```diff
@@ -0,0 +1,536 @@
+        if (type == GameType.BOT) {
+            painter.setOpponentName("Bot");
+            bot = new Bot();
+            bot.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+        } else {
+            painter.setOpponentName(NetworkController.getOpponentName());
+        }
+        generation = new ManaGenerator();
+        generation.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+    }
+
+    public void endGame() {
+        isStopped = true;
+        if (bot != null) {
+            bot.stop();
+        }
+        generation.stop();
+    }
+
+    /** Method activating spell from the player. */
+    public void playerSpell(String spell) {
+        String ability = logic.ableToThrowTheSpell(spell);
+        if (ability != "ok") {
+            painter.sendNotification(ability);
+            return;
+        } else {
+            logic.initializeCast(spell);
+        }
+        painter.lockInput();
+        painter.timerCast(logic.getCastByName(spell));
+        if (type == GameType.MULTIPLAYER) {
+            NetworkController.sendSpell(logic.getIDByName(spell));
+        }
+        ThrowPlayerSpell throwPlayerSpell = new ThrowPlayerSpell();
+        throwPlayerSpell.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, spell);
+    }
+
+    /** Method activating spell from the opponent. */
+    public void opponentSpell(int spellID) {
+        if (isStopped) {
+            return;
+        }
+        painter.showOpponentSpell(logic.getNameById(spellID));
+        ThrowOpponentSpell throwOpponentSpell = new ThrowOpponentSpell();
+        throwOpponentSpell.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR, logic.getNameById(spellID));
+    }
+
+    private void generateMana(int mana) {
+        logic.generateMana(mana);
+        painter.setPlayerMP(logic.getPlayerMP());
+    }
+
+    /** Part of the opponent spell after cast. */
+    private void throwOpponentSpell(String spell) {
+        if (isStopped) {
+            return;
+        }
+        painter.hideOpponentSpell();
+        showOpponentCast(spell);
+        logic.opponentSpell(spell);
+        painter.setPlayerHP(logic.getPlayerHP());
+        if (logic.getPlayerHP() == 0) {
+            endGame();
+            painter.endGame(GameResult.LOSE);
+        }
+    }
+    /** Part of the player spell after cast. */
+    private void throwPlayerSpell(String spell) {
+        painter.unlockInput();
+        if (isStopped) {
+            return;
+        }
+        showPlayerCast(spell);
+        logic.playerSpell(spell);
+        painter.setPlayerMP(logic.getPlayerMP());
+        painter.setPlayerHP(logic.playerHP);
+        painter.setOpponentHP(logic.getOpponentHP());
+        if (logic.getOpponentHP() == 0) {
+            endGame();
+            painter.endGame(GameResult.WIN);
+        }
+    }
+
+    /** Part of the player spell after cast and throwing. */
+    private  void stopPlayerSpell(String spell) {
+        if (logic.getTypeByName(spell).equals("buff")) {
```

Comment:
The problem is that this code calls method `getTypeByName()` what works with the database and receives string as a result of query. So if we want to use enums instead of strings, we should have a code for matching spell types and query result. This code would be called every time  when something happens in the game. Is it really necessary? Or maybe the problem I have described is not important?

Reply:
I would rather convert to an Enum as soon as the string is extracted from the database -- and use the enum everywhere else. Conversely, convert to string just before writing to the db.